### ISSUE TITLE
ref ternary expressions should not merge nullability from unreachable branches

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#41908]
**Version Used**: 
```
commit 7aa825c699587a792395b75831109fff2deab183 (HEAD -> master, origin/master, origin/HEAD)
Merge: c288341698 e16f4ede7b
Author: Manish Vasani <mavasani@microsoft.com>
Date:   Mon Feb 24 05:20:55 2020 -0800

    Merge pull request #41635 from mavasani/UnusedParameter

    Remove an unused parameter
```

**Steps to Reproduce**:

Compile the following code
```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;

class C<T>
{
  [DoesNotReturn] public ref string? Fail() => throw null!;
  
  public void M1(string s)
  {
    string y = "".Length > 0 ? Fail() : s;
    ref string x = ref "".Length > 0 ? ref Fail() : ref s;
  }
}
```

**Expected Behavior**:
No warnings. There's only one valid branch in both conditional expressions

**Actual Behavior**:
`warning CS8619: Nullability of reference types in value of type 'string?' doesn't match target type 'string'.` is reported for the ref conditional expression only.

**Notes**
Since it's impossible for the conditional expression to return `string?` as the method is annotated as `[DoesNotReturn]` it should neither produce warnings for converting `string?` to the conditional expression's common type nor take its top level nullability into account when inferring nullability of the conditional expression. Note that this is how Roslyn behaves with regular (non-ref) conditional expression as it recognizes that `"".Length > 0 ? Fail() : s` returns non-nullable string and does not warn about converting it to non-nullable variable (such warnings appear if you remove the `[DoesNotReturn]` annotation

**Comment by olivier-spinelli** [comment]
The switch expression suffers from this issue also (typically the case of the ultimate `_` fallback).
But, unless I missed something, this is a much more general (and IMO annoying) issue. The following (totally stupid but basically imperative) code cannot compile:
```csharp
[DoesNotReturn]
public static void Bang() => throw null!;

static string DoesNotReturn_is_ignored( string input )
{
    string r;
    if( input == null ) Bang();
    else r = input;
    return r; // Compiler Error CS0165: Use of unassigned local variable 'r'
}
```
